operator_compile_check v0

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #103198
* #103197
* #103196

This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

zou3519 committed:
operator_compile_check v0

This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/optests/compile_check.py:
```diff
@@ -0,0 +1,98 @@
+        # We should change this when Dynamo does support tree_map.
+        # Note that what we have is still correct, because operators do
+        # not accept nested lists.
+        args = tuple(clone_arg(a) if isinstance(a, torch.Tensor) else a for a in args)
```
> Note that what we have is still correct, because operators do not accept nested lists

This example will break if you have a custom op that takes in an ordinary TensorList, right? Is that explicitly not supported today either?

zou3519 replied:
This code should work for TensorList. `clone_arg` supports cloning Tensors or TensorList and is called on each of the args to the function. I can add a test case to demonstrate.

zou3519 replied:
Actually, we already have a test case: numpy_cat in the custom_op_db has TensorList inputs: https://github.com/pytorch/pytorch/blob/7108c035bc0309f60fc86d32a42335b8808576f9/torch/testing/_internal/custom_op_db.py#L209

zou3519 replied:
Oh, yes, you're right. I'll delete the `isinstance(a, torch.Tensor) ` check

bdhirsh commented on torch/testing/_internal/optests/compile_check.py:
```diff
@@ -0,0 +1,98 @@
+
+    # Add some clones to the beginning of func so that it is actually out-of-place.
+    # A lot of our tests (e.g. aot_autograd_check) assume that the function passed
+    # in is out-of-place and contains a call to the operator in being tested.
```
mostly as a confirmation/reminder for me: `CustomOp` doesn't allow you to create ops with mutable schemas, so for `CustomOp`, these clones shouldn't be necessary. Is this API also meant to be used with existing C++ custom ops, which might perform views and mutations?

zou3519 replied:
> Is this API also meant to be used with existing C++ custom ops, which might perform views and mutations?

Yes

bdhirsh replied:
Hmm - totally reasonable if you think this isn't worth doing, but: given that AOTAutograd has to do a whole bunch of extra work when mutation and aliasing comes into the picture, do you think it's worth exercising that logic more directly in the custom op tests?

Specifically: Written as is (where we clone all tensor inputs), we'll never end up invoking AOTAutograd's custom logic around "what happens when you do an input mutation, or return an output that aliases an input". `test/functorch/test_aotdispatch.py` has a bunch of tests that try to explicitly handle this case, by:

(1) running functions that cause input mutations, and checking that the inputs match compared to eager

(2) run functions that return aliases of inputs, explicitly mutate the output afterwards, and check that inputs match (and also confirm that the computed gradients match

One argument for *not* doing this that I think is fair: If that extra logic to fail, that would be more indicative of an AOTAutograd bug, than a bug in your custom op implementation. Should `compile_check()` be playing that role?

Then again, we're also running inductor directly as part of this check, so maybe we do just want to exercise as much of the system as possible.

bdhirsh replied:
"custom C++ op that returns an alias of an input, but has a custom autograd rule" is probably a good example of something that this could catch (aka https://github.com/pytorch/pytorch/issues/102970)

zou3519 replied:
Is this something you think we would/should add to all of the OpInfo tests for AOTAutograd?

bdhirsh replied:
hmm. I think it probably isn't worth adding to all the OpInfo tests, since this doesn't matter for 99% of ops (and that would probably increase test time quite a bit). I guess I'm just a bit more paranoid about custom ops, since we've seen custom ops with wilder schemas than most of our ATen ops.

zou3519 replied:
IMO it's good to add more tests. Would it be fine if I took this as a follow-up item? compile_check as written is useful today, not public yet, and figuring out where to check input mutations is a bit complicated

bdhirsh replied:
totally fine!

zou3519 commented:
NB: The dynamo failures are Dynamo not handling CustomOp correctly. This is a pre-existing problem and was masked because test_python_dispatch (where the rest of the CustomOp tests are) are in the dynamo testing denylist. I'll likely just add xfails for dynamo.

soulitzer commented on test/test_custom_op_testing.py:
```diff
@@ -0,0 +1,327 @@
+            return op(x)
+
+        x = torch.tensor(3.14159 / 3, requires_grad=True)
+        with self.assertRaisesRegex(AssertionError, "not completely traceable"):
```
If dynamo is able to guard on this and trigger a recompile, is that technically considered "traceable" in terms of torch.compile? 

I think the error message is already pretty clear that we're talking about make_fx here and that it is due to depending global state, but maybe its also worth explaining that even though you failed this test, your function may still work with torch.compile?

zou3519 replied:
Yes, I'll make it clearer. Ideally what the user provides to operator_compile_check is a single operator (instead of an operator + some more calls).

bdhirsh reviewed (approved):
🚀

soulitzer commented on test/test_custom_op_testing.py:
```diff
@@ -0,0 +1,327 @@
+            return x.view_as(x)
+
+        def foo_meta(x):
+            return x.view_as(x)
```
Just for my understanding, if everything else is correct, does it matter whether the meta has the view-ness as the impl?

bdhirsh replied:
yep! Especially for functionalization - we rely on the meta functions when we're tracing with functionalization, so functionalization can know what aliases to replay any mutations on.

soulitzer replied:
I see, thanks!

soulitzer commented on test/test_custom_op_testing.py:
```diff
@@ -0,0 +1,327 @@
+        x = torch.tensor([0, 1.], requires_grad=True)
+        with self.assertRaisesRegex(
+                RuntimeError,
+                'FooBackward returned an invalid gradient at index 0'):
```
Just curious, why is this the error message for this failure?

zou3519 replied:
AOTAutograd runs Foo's forward and backward.
On the forward, it accepts a tensor of shape [2] and returns a tensor of shape [1, 2].
On the backward, it accepts a tensor of shape [1, 2] and returns a Tensor of shape [1, 2]. Autograd ends up yelling.

We should have a dedicated test to check the FakeTensor rule (it's not in this PR because I'm not sure how to actually create FakeTensors in dynamic mode). I'll add that in a follow up.

soulitzer reviewed (approved):
LGTM, just some minor questions

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 committed:
Update on "operator_compile_check v0"


This PR adds `operator_compile_check` (pls bikeshed name), a gradcheck-like
API to test if a custom operator is supported by torch.compile.

The API is scoped to check only that the interaction between the
operator and torch.compile works (e.g. it is not going to include
gradcheck). Concretely, it currently checks the following things:
- schema correctness
- make_fx traceable (static shapes)
- aot_autograd correctness (static shapes)
- torch.compile correctness, with and without inductor (static shapes)
- make_fx traceable (dynamic shapes)
- aot_autograd correctness (dynamic shapes)
- torch.compile correctness, with and without inductor (dynamic shapes)

Test Plan:

We test a bunch of error cases, including many failure modes that have tripped
us up in the past, and assert that they (mostly) have nice error messages:
- incorrect schema (mutates)
- incorrect schema (has a view)
- missing abstract impl
- incorrect abstract impl
- missing functionalization kernel
- autograd registered at CPU/CUDA keys
- operator is not traceable

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge
